achievements.py

```diff
@@ -40,3 +41,48 @@ def SymmetricCoffee(consumption):
         new = Achievement(description=get_description_for_key(key), key=key)
         consumption.user.achievements.append(new)
         consumption.user.save()
+
+@Consumption.achievement_function
+def Minimalist(consumption, debug=True):
+    key = 'minimalist'
+    n_consumptions_total = len(Consumption.objects(user=consumption.user))
+
+    if n_consumptions_total < 5:
+        return
+
+    today = consumption.date
+    p1, p2 = [p[0] for p in COFFEE_PRICES]
+
+    last_consumption_date = Consumption.objects(
+            user=consumption.user
+        ).order_by('-date').first().date
+
+    if not last_consumption_date.weekday() == 4:
+        return
+
+    if last_consumption_date.date() == today.date():
+        return
+
+    shift = 1 if n_consumptions_total==5 else 0
+    if (Consumption.objects(user=consumption.user).order_by('-date')[4-shift].date.date() == \
+        Consumption.objects(user=consumption.user).order_by('-date')[5-shift].date.date()):
+        return
+
+    complete_week_deltas = [
+        datetime.timedelta(i) for i in range(0,5,1)
+    ]
+
+    consumptions_to_check = Consumption.objects(
```

Comment:
Try to cache the Consumption objects to save some db hits. I think you could get along with only one `Consumption.objects(user=consumption.user)` query

Reply:
Is there a common practice to cache those things? What I can think of is looping over this 5/6 consumptions and store the dates/consumptions in a dictionary/two lists and apply the checks on theses.

Reply:
Since you’re only interested in consumptions that are maximum 7 days old you could get a list of those once and work with copies of that.

I think you could even write a single aggregate to solve most of the logic in one go. Something like

1. match consume from given User younger than one week
2. group total consumptions by day
3. check that there are 5 entries and a sum of 5 consumptions

Bam! 💥